### PR TITLE
Prefs[TASK[28593] Handle events (Hotel Tab)

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 28593: Handle events (Hotel Tab)](https://edu-redmine.sun-asterisk.vn/issues/28593)

## WHAT
- Handle events (Hotel Tab)
## HOW
- Handle events (Hotel Tab)
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/MaterialPickDialogFragment.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.sunasterisk.travelapp.ui.custom
+
+import android.app.Dialog
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import androidx.fragment.app.DialogFragment
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
+
+open class MaterialPickDialogFragment : DialogFragment() {
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/MaterialPickDialogFragment.kt]: có lý do gì class này open không nhỉ?

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 2]
```diff
@@ -1,27 +1,53 @@
...
 import com.google.android.material.bottomsheet.BottomSheetBehavior
+import com.google.android.material.datepicker.MaterialDatePicker.Builder.dateRangePicker
 import com.sunasterisk.travelapp.R
 import com.sunasterisk.travelapp.ui.base.BaseFragment
 import com.sunasterisk.travelapp.utils.location
+import com.sunasterisk.travelapp.utils.screenHeight
 import kotlinx.android.synthetic.main.fragment_hotel_tab.*
+import java.util.*
+
+typealias MyPair = androidx.core.util.Pair<Long, Long>
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]: Dùng có 1 lần thì không cần khai báo kiểu đâu ông. Dùng luôn Pair<>...

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 3]
```diff
@@ -1,27 +1,53 @@
...
+            activity?.supportFragmentManager?.let {
+                picker.show(it, picker.toString())
+            }
+            picker.addOnPositiveButtonClickListener {
+                textViewFilterDates.text = picker.headerText
+            }
+        }
+
+        textViewFilterTravelers.setOnClickListener {
+            fragmentManager?.let { manager -> TravelersPickDialogFragment().show(manager, "") }
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]:
```suggestion
            fragmentManager?.let { manager -> TravelersPickDialogFragment().show(manager, TravelersPickDialogFragment::class.java.simpleName) }
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 4]
```diff
@@ -1,27 +1,53 @@
...
+
+    private fun setClickListener() {
+        val builder = dateRangePicker()
+        textViewFilterDates.setOnClickListener {
+            val now = Calendar.getInstance()
+            builder.setTheme(R.style.ThemeOverlay_Catalog_MaterialCalendar_Custom)
+            builder.setSelection(MyPair(now.timeInMillis, now.timeInMillis))
+            val picker = builder.build()
+            activity?.supportFragmentManager?.let {
+                picker.show(it, picker.toString())
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]: Đoạn này đang lấn cấn cái trách nhiệm của hàm nha. hàm setClick thì chỉ handle click thôi, ông nên tách rành mạch ra setupViews() và handleEvents() nha

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/TravelersPickDialogFragment.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import com.sunasterisk.travelapp.R
+import com.sunasterisk.travelapp.ui.custom.MaterialPickDialogFragment
+
+class TravelersPickDialogFragment : MaterialPickDialogFragment() {
+
+    override fun onCreateView(inflater: LayoutInflater,
+                              container: ViewGroup?,
+                              savedInstanceState: Bundle?): View? {
+        return inflater.inflate(R.layout.fragment_travelers_pick_dialog, container, false)
+    }
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/TravelersPickDialogFragment.kt]:
```suggestion
                              savedInstanceState: Bundle?): View? =
        inflater.inflate(R.layout.fragment_travelers_pick_dialog, container, false)
```

File: app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt [hunk 6]
```diff
@@ -53,3 +53,11 @@ fun Context.showToast(obj: Any) {
     }
     Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
 }
+
+fun Context.screenHeight(): Int? {
+    return this.resources?.displayMetrics?.heightPixels
+}
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt]:
```suggestion
fun Context.screenHeight(): Int? = this.resources?.displayMetrics?.heightPixels
```

File: app/src/main/java/com/sunasterisk/travelapp/utils/ViewExt.kt [hunk 7]
```diff
@@ -20,6 +21,12 @@ fun View.gone() {
     visibility = View.GONE
 }
 
+fun Long.calendar(): Calendar {
+    return Calendar.getInstance().also {
+        it.time = Date(this)
+    }
+}
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/utils/ViewExt.kt]:
```suggestion
fun Long.calendar() = Calendar.getInstance().also {it.time = Date(this)}
```

File: app/src/main/res/layout/fragment_hotel_tab.xml [hunk 8]
```diff
@@ -41,16 +41,17 @@
...
+            android:id="@+id/textViewFilterDates"
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:background="@android:color/transparent"
             android:drawableStart="@drawable/ic_dates"
             android:drawablePadding="@dimen/dp_8"
             android:hint="@string/title_select_dates"
             android:padding="@dimen/dp_12"
-            android:paddingLeft="@dimen/dp_8" />
+            android:paddingLeft="@dimen/dp_8"
```
Comment (tuanpa-0908) [inline app/src/main/res/layout/fragment_hotel_tab.xml:55]: không add paddingRight là sẽ có lint warning nha

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/MaterialPickDialogFragment.kt (hunk 1 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/custom/MaterialPickDialogFragment.kt]: Ok ông, để tái sử dụng cái custom dialog fragment, vì về sau có thể có nhiều kiểu dialog khác nhau, chỉ cần set layout view cho mỗi dialog đấy thôi, ko cần viết lại phần Builder

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt (hunk 2 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]: ok ông

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 9]
```diff
@@ -30,4 +54,16 @@ class HotelTabFragment private constructor() : BaseFragment() {
         const val TITLE = "Sleep"
         fun newInstance() = HotelTabFragment()
     }
+
+    override fun onClick(view: View?) {
+        when (view?.id) {
+            textViewFilterDates.id -> setupDatePickerBuilder()
+            textViewFilterTravelers.id -> {
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]:
```suggestion
            R.id.textViewFilterDates -> setupDatePickerBuilder()
            R.id.textViewFilterTravelers-> {
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt [hunk 10]
```diff
@@ -30,4 +54,16 @@ class HotelTabFragment private constructor() : BaseFragment() {
...
+                fragmentManager?.let { manager ->
+                    TravelersPickDialogFragment().show(manager,
+                            TravelersPickDialogFragment::class.simpleName)
+                }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]: conventions
https://kotlinlang.org/docs/reference/coding-conventions.html#method-call-formatting

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/TravelersPickDialogFragment.kt [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+import android.view.ViewGroup
+import com.sunasterisk.travelapp.R
+import com.sunasterisk.travelapp.ui.custom.MaterialPickDialogFragment
+
+class TravelersPickDialogFragment : MaterialPickDialogFragment() {
+
+    override fun onCreateView(inflater: LayoutInflater,
+                              container: ViewGroup?,
+                              savedInstanceState: Bundle?) = inflater.inflate(R.layout.fragment_travelers_pick_dialog, container, false)
+}
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/TravelersPickDialogFragment.kt]: fucntion formating 
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt [hunk 12]
```diff
@@ -53,3 +53,8 @@ fun Context.showToast(obj: Any) {
     }
     Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
 }
+
+fun Context.screenHeight(): Int? = this.resources?.displayMetrics?.heightPixels
+
+
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt]: double empty line

File: app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt [hunk 13]
```diff
@@ -53,3 +53,8 @@ fun Context.showToast(obj: Any) {
     }
     Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
 }
+
+fun Context.screenHeight(): Int? = this.resources?.displayMetrics?.heightPixels
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt]:
```suggestion
fun Context.screenHeight() = resources?.displayMetrics?.heightPixels?:0
```

File: app/src/main/res/values/dimens.xml [hunk 14]
```diff
@@ -11,4 +11,12 @@
...
     <dimen name="dp_72">72dp</dimen>
     <dimen name="dp_172">172dp</dimen>
+    <dimen name="dp_2">2dp</dimen>
+    <dimen name="dp_32">32dp</dimen>
+    <dimen name="dp_128">128dp</dimen>
+    <dimen name="dp_48">48dp</dimen>
+    <dimen name="dp_96">96dp</dimen>
+    <dimen name="dp_42">42dp</dimen>
+    <dimen name="dp_25">25dp</dimen>
+    <dimen name="dp_6">6dp</dimen>
```
Comment (huytq-2422) [inline app/src/main/res/values/dimens.xml]: sắp xếp theo thứ tự nào

File: app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt (hunk 10 above)
Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/travelapp/ui/home/hotel/HotelTabFragment.kt]:
```suggestion
                val dialog = TravelersPickDialogFragment()
                fragmentManager?.let { 
                    dialog.show(it, TravelersPickDialogFragment::class.simpleName)
                }
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/MaterialPickDialogFragment.kt [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+            dialogView?.let {
+                onViewCreated(it, savedInstanceState)
+            }
+            setView(dialogView)
+        }.create()
+    }
+
+    override fun getView(): View? {
+        return dialogView
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/MaterialPickDialogFragment.kt]: https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt [hunk 16]
```diff
@@ -53,3 +53,7 @@ fun Context.showToast(obj: Any) {
     }
     Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
 }
+
+fun Context.screenHeight() = this.resources?.displayMetrics?.heightPixels ?: 0
+
+fun Context.screenWidth(): Int? = this.resources?.displayMetrics?.widthPixels ?: 0
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/utils/AppCompatActivityExt.kt]:
```suggestion
fun Context.screenWidth() =resources?.displayMetrics?.widthPixels ?: 0
```

**Pull request merged**